### PR TITLE
[Feature] Support ViTPose

### PR DESCRIPTION
**Pull request by Annbless**
## Motivation

Merge the ViTPose variants code and pre-trained models into mmpose.

## Modification

1. Add a vit backbone model in mmpose/models/backbones. The ```__init__``` file is modified accordingly.
2. Add the config files and corresponding markdown files in the configs folder.
3. Fix a bug in the registration of layer-wise learning rate decay.
4. Add a 'resize_upsample4' input transformation in the mmpose/models/heads/topdown_heatmap_simple_head.py file to support the simple decoder in ViTPose. It has no influence on other models. 

## BC-breaking (Optional)

No.

## Checklist

**Before PR**:

- [X] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [X] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [X] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [X] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [X] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by ly015** [comment]
Thank you very much for your help! For now, there are lint issues in the code. Could you please install pre-commit hooks (see our [docs](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md)) and run ``pre-commit run --all-files`` in your local repo? The lint issues will be fixed automatically.

**Comment by Annbless** [comment]
Thanks for your instruction. The code is checked now. How can I upload the pre-trained weights and logs for the ViTPose variants? Can I provide the links using OneDrive or Google Drive?

**Comment by jin-s13** [comment]
Thanks. Both OneDrive and Google Drive are welcome. 

BTW, would you mind adding some unit-tests? An example could be found https://github.com/open-mmlab/mmpose/pull/1907/files#diff-dadc2075341a40335f28131ceaf3d0d415e5c316c54bb6b0a0741aeb002db24e

**Comment by ly015** [comment]
The unit test of ViTPose seems failed. For quick debugging, you can run unit tests locally by ``$pytest tests/``.

**Comment by Annbless** [comment]
Thanks a lot for your help! The files and configs have been updated now. The pre-trained models and logs are available at [Onedrive](https://1drv.ms/u/s!AimBgYV7JjTlgccv_ney00z-K9TAGQ?e=fQIIAB). The uncovered codes by the unit test are mostly existing in the init weight using the pre-trained models part (for example, changing the tensor name between the MAE pre-trained models and the backbones). We are wondering how can we cover these parts in the unit test. Can we produce pseudo checkpoints via ```torch.save``` in the unit test to cover the rename parts? We have tested this parts via re-training the models for several epochs and find them work well.

**Comment by Annbless** [comment]
We also fixed some bugs caused by the updated NumPy version in the dataset files. Please check the recent commits. By the way, it seems that the current failed build is cased by the HTTP error.

**Comment by Annbless** [comment]
Hi @ly015, would you mind restarting the failed checks? I just checked the logs, and it seems that the pip installation caused the error. Thanks a lot.

**Comment by Annbless** [comment]
It seems that the current failure information is related to the docker version... Should I open a new PR for the dev-1.x branch and close this PR instead? Thanks a lot for your patience.

**Comment by Annbless** [comment]
Hi @ly015, it seems that the current failure case is in loading the video in ```test_inference.py```, where no frames are detected after the [command](https://github.com/open-mmlab/mmpose/actions/runs/4122842964/jobs/7120120416#step:12:352).
To this end, is there anything we can do to aid the merging?

Thanks a lot.

**Comment by ly015** [comment]
We will help check and fix the CI problem.

**Comment by Annbless** [comment]
Hi, is there anything we can do to help fix the CI problem? Besides, could we open a new PR based on the dev-1.x branch to merge the ViTPose variants into mmpose? Thanks for your response.

File: mmpose/models/backbones/vit.py [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+                                                  'attn.attn.in_proj_')
+                    elif 'attn.proj' in new_key:
+                        new_key = new_key.replace('attn.proj',
+                                                  'attn.attn.out_proj')
+                    new_ckpt[new_key] = v
+                elif k.startswith('patch_embed'):
+                    new_key = k.replace('patch_embed.proj',
+                                        'patch_embed.projection')
+                    new_ckpt[new_key] = v
+                else:
```
Comment (LareinaM) [inline mmpose/models/backbones/vit.py:300]: The downloaded backbone has keys `norm.weight` and `norm.bias`, but in the model the two are called `last_norm.weight` and `last_norm.bias`, it is necessary to add another conversion?

Comment (Annbless) [reply mmpose/models/backbones/vit.py:300]: No need. The last norm layer is re-initialized for the pose dataset.

**Comment by LareinaM** [comment]
I have trained these models using your code and downloaded pretrained backbones. However, the results for some models mismatch with your record.

> With classic decoder

| Arch | Input Size | AP | AP<sup>50</sup> | AP<sup>75</sup> | AR | AR<sup>50</sup> | 
| :-------------------------------------------- | :--------: |  :-------------: | :-------------: | :---: | :-------------: | :-------------------------------------------: | 
| ViTPose-S | 256x192 | 0.737 | 0.905 | 0.813 | 0.790 | 0.943 | 
| ViTPose-B | 256x192 | 0.751 | 0.905 | 0.823 | 0.803 | 0.944 | 
| ViTPose-L | 256x192 | 0.777 | 0.915 | 0.850 | 0.828 |  0.953 | 
| ViTPose-H | 256x192 | 0.785 | 0.914 | 0.853 | 0.835 | 0.951 |

> With simple decoder 

| Arch | Input Size | AP | AP<sup>50</sup> | AP<sup>75</sup> | AR | AR<sup>50</sup> |
| :-------------------------------------------- | :--------: | :---: | :-------------: | :-------------: | :---: | :-------------: |
| ViTPose-S | 256x192 | 0.736 | 0.904 | 0.812 | 0.790 | 0.942 |
| ViTPose-B | 256x192 | 0.750 | 0.905 | 0.827 | 0.805 | 0.944 | 
| ViTPose-L | 256x192 | 0.774 | 0.911 | 0.847 | 0.826 | 0.950 | 
| ViTPose-H | 256x192 | 0.785 | 0.915 | 0.854 | 0.835 | 0.952 | 

The validation accuracy is the same for all models.

I also noticed that the highest accuracy for large and huge models is at around 80th epoch. Maybe there is a problem with the optimizer? Have you validated the training process on this PR?

**Comment by Annbless** [comment]
Hi 
We have re-trained the models and figured out that the performance drop is caused by the difference between the transformer layers implemented by mmcv and timm. Is it possible for us to use timm for the backbone implementation?

**Comment by ly015** [comment]
Yes, you can use timm for the backbone implementation. There is a tutorial in MMDetection on how to use timm backbones in MMDetection through an MMClassification wrapper, which should also be applicable for MMPose: https://mmdetection.readthedocs.io/en/latest/tutorials/how_to.html#use-backbone-network-in-timm-through-mmclassification

The above tutorial is just for your reference. You can use any approach to integrate timm backbones in your implementation.

**Comment by Annbless** [comment]
Hi there,

Thanks for your patience. We have uploaded a timm version of the ViTPose.

The training logs are available here.
[vitpose_base.log](https://github.com/open-mmlab/mmpose/files/11046447/vitpose_base.log)
[vitpose_simple_base.log](https://github.com/open-mmlab/mmpose/files/11046449/vitpose_simple_base.log)
[vitpose_small.log](https://github.com/open-mmlab/mmpose/files/11046451/vitpose_small.log)
[vitpose_simple_small.log](https://github.com/open-mmlab/mmpose/files/11046450/vitpose_simple_small.log)

**Comment by Annbless** [comment]
Hi there,

Are there any things we could help to aid the merge of the PR? We are willing to provide more information.

Best,
